### PR TITLE
Widget custom selector

### LINKED ISSUE #7516: Adds custom layer's selector for new widgets
screen shot 2016-05-25 at 18 36 32

### PR DESCRIPTION
**Pull request by nobuti**
This PR fixes #7516.

add-widget

@piensaenpixel the styles are pretty basic, let me know what you want me to improve please.

cc @xavijam

### CODE REVIEW

File: app/assets/stylesheets/editor-3/widget-list.css.scss [hunk 1]
```diff
@@ -39,13 +39,28 @@
 }
 .WidgetList-inner {
   @include display-flex;
+  @include flex-direction(column);
   flex-wrap: wrap;
   width: 100%;
```
Comment (piensaenpixel) [inline app/assets/stylesheets/editor-3/widget-list.css.scss]: i think we only need the width
image

File: app/assets/stylesheets/editor-3/widget-list.css.scss [hunk 2]
```diff
@@ -39,13 +39,28 @@
...
 }
 .WidgetList-title {
   @include display-flex;
 }
-
+.Widget-selectorLayer {
+  position: relative;
+  background-color: #F9F9F9;
+  display: inline-block;
+  min-width: 200px;
```
Comment (piensaenpixel) [inline app/assets/stylesheets/editor-3/widget-list.css.scss]: i think we dont need display and min-width: 
image

File: app/assets/stylesheets/editor-3/widget-list.css.scss [hunk 3]
```diff
@@ -39,13 +39,28 @@
...
+}
+.Widget-selectorLetter {
+  border-radius: 4px;
+  background-color: #F15743;
+  color: #fff;
```
Comment (piensaenpixel) [inline app/assets/stylesheets/editor-3/widget-list.css.scss]: use $cWhite

File: app/assets/stylesheets/editor-3/widget-list.css.scss [hunk 4]
```diff
@@ -39,13 +39,28 @@
...
-
+.Widget-selectorLayer {
+  position: relative;
+  background-color: #F9F9F9;
+  display: inline-block;
+  min-width: 200px;
+}
+.Widget-selectorLetter {
+  border-radius: 4px;
+  background-color: #F15743;
```
Comment (piensaenpixel) [inline app/assets/stylesheets/editor-3/widget-list.css.scss]: I suppose here the color should be in the html and change depending the principal color layer

File: app/assets/stylesheets/editor-3/widget-list.css.scss [hunk 5]
```diff
@@ -39,13 +39,28 @@
...
 }
-
+.Widget-selectorLayer {
+  position: relative;
+  background-color: #F9F9F9;
+  display: inline-block;
+  min-width: 200px;
+}
+.Widget-selectorLetter {
+  border-radius: 4px;
```
Comment (piensaenpixel) [inline app/assets/stylesheets/editor-3/widget-list.css.scss]: maybe use $halfBaseSize var

File: app/assets/stylesheets/editor-3/widget-list.css.scss [hunk 6]
```diff
@@ -39,13 +39,28 @@
...
+  background-color: #F15743;
+  color: #fff;
+  font-size: 10px;
```
Comment (piensaenpixel) [inline app/assets/stylesheets/editor-3/widget-list.css.scss]: use CDB-Text CDB-Size-small and u-whiteTextColor helpers

File: app/assets/stylesheets/editor-3/widget-list.css.scss [hunk 7]
```diff
@@ -39,13 +39,28 @@
...
+  text-transform: uppercase;
+  padding: 1px 6px;
+  display: inline-block;
```
Comment (piensaenpixel) [inline app/assets/stylesheets/editor-3/widget-list.css.scss]: what do you think if we remove this one and use u-flex in the parent?

File: lib/assets/javascripts/cartodb3/components/modals/add-widgets/layer-selector-view.js [hunk 8]
```diff
@@ -20,25 +22,57 @@ module.exports = cdb.core.View.extend({
...
-      layerIndex: i,
-      layerNames: _.map(tuples, function (item) {
-        return item.analysisDefinitionModel.get('id');
-      })
+  _buildOptions: function () {
+    return _.map(this.model.get('tuples'), function (item, index) {
+      return {
+        val: index,
+        label: item.analysisDefinitionModel.get('id'),
+        color: '#ff6600' // To be changed item.layerDefinitionModel.get('color')
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/modals/add-widgets/layer-selector-view.js]: cc @alonsogarciapablo :grin:

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/modals/add-widgets/layer-selector-view.js]: Btw, could we add that property, although they will appear with the same color for the moment?

Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/components/modals/add-widgets/layer-selector-view.js]: We could use `item.layerDefinitionModel.get('color')` here and default that property to `#ff6600` in layerDefinitionModel...

File: lib/assets/javascripts/cartodb3/components/form-components/editors/select/select-layer-list-item-view.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    'click': '_onClick'
+  },
+
+  initialize: function (opts) {
+    this.options = _.extend({}, this.options, opts);
+  },
+
+  render: function () {
+    this.$el.empty();
+    this.clearSubViews();
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/form-components/editors/select/select-layer-list-item-view.js]: I feel this thing should be before "emptying" the HTML content.

File: lib/assets/javascripts/cartodb3/components/form-components/editors/select/select-layer-list-item-view.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  tagName: 'li',
+
+  events: {
+    'mouseenter': '_onMouseEnter',
+    'mouseleave': '_onMouseLeave',
+    'click': '_onClick'
+  },
+
+  initialize: function (opts) {
+    this.options = _.extend({}, this.options, opts);
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/form-components/editors/select/select-layer-list-item-view.js]: We don't need this, cdb.core.View does that internally (in the constructor).

**Comment by xavijam** [comment]
Just minor comments, ready to be :ship: :it:

File: app/assets/stylesheets/editor-3/layer-selector.css.scss [hunk 11]
```diff
@@ -0,0 +1,9 @@
+.SelectorLayer {
+  position: relative;
+  background-color: #F9F9F9;
+}
+.SelectorLayer-letter {
+  padding: 1px 6px;
+  border-radius: 4px;
```
Comment (piensaenpixel) [inline app/assets/stylesheets/editor-3/layer-selector.css.scss]: could we use $halfBaseSize?

File: app/assets/stylesheets/editor-3/layer-selector.css.scss [hunk 12]
```diff
@@ -0,0 +1,9 @@
...
+  text-transform: uppercase;
```
Comment (piensaenpixel) [inline app/assets/stylesheets/editor-3/layer-selector.css.scss]: Helper? u-upperCase

File: app/assets/stylesheets/editor-3/layer-selector.css.scss [hunk 13]
```diff
@@ -0,0 +1,9 @@
+.SelectorLayer {
+  position: relative;
+  background-color: #F9F9F9;
```
Comment (piensaenpixel) [inline app/assets/stylesheets/editor-3/layer-selector.css.scss]: use $cThirdBackground

**Comment by nobuti** [comment]
🍏  🚀

**Pull request merged**